### PR TITLE
Assorted refactorings

### PR DESCRIPTION
**Pull request by iaindillingham**
Assorted refactorings of the redaction, crosstab, and import functionality. Draft, at present, because I'd like to finish by refactoring `find_save_tools.make_folders`.

I'm trying to go step-by-step with these commits. First, working on tests; then on docstrings; then on arguments and signatures; and finally the functions themselves. However, I'm also mindful of addressing the following issues, either in the PR or in future refactorings:

* Running the tests shouldn't write files to the project directory
* Adding assertions to all tests, unless explicitly stated
* Replacing fixture files with fixture functions/factory functions
* Sticking to a single docstring style, without type hints in the docstrings themselves
* Typing all functions with the exception of protected (`_*`) and private (`__*`) functions
* Adding return type annotations
* Simplifying confusing module names
* Consistent variable names -- avoiding `df` for Pandas data frames

### CODE REVIEW

File: action/find_save_tools.py [hunk 1]
```diff
@@ -1,52 +1,46 @@
...
     else:
-        raise ImportActionError("Unsupported filetype attempted to be imported")
+        raise ImportActionError(f"'{file_path.suffix}' is not a supported file-type")
 
-    # checks that the variables defined in the json are column names in the
-    # csv and raises an error if note
-    for table_names, instructions in variable_json.items():
-        if not set(instructions["variables"]).issubset(df.columns.values):
-            raise ImportActionError
+    variables = set(itertools.chain(*[x["variables"] for x in table_configs.values()]))
```
Comment (iaindillingham) [inline action/find_save_tools.py:26]: `itertools` is powerful and terse: hopefully you understand what's going on here. Working from the inside out:

* Iterate over each item in `table_configs.values()`, which is a list (more-or-less 🙂), returning the contents of the `variables` property, which is a list. This gives us a list of lists.
* Expand (`*`) the list of lists. [`itertools.chain`](https://docs.python.org/3.8/library/itertools.html#itertools.chain) receives multiple lists as multiple positional arguments. It iterates over elements from the first list until this list is exhausted; then the second list; then the third list; etc.
* Reduce the list of variables to a set of variables (i.e. no duplicates)

File: action/find_save_tools.py [hunk 2]
```diff
@@ -1,52 +1,46 @@
...
+    if not variables <= set(table.columns):
```
Comment (iaindillingham) [inline action/find_save_tools.py:27]: `<=` or [`issubset`](https://docs.python.org/3.8/library/stdtypes.html#frozenset.issubset). It's important that this isn't a _proper_ subset, because they might be identical.

File: action/find_save_tools.py [hunk 3]
```diff
@@ -1,52 +1,46 @@
...
-    # returns the csv
-    return df
+    return table
 
 
-def make_folders(table_config_json, path=None):
-    """
-    Makes the output folders for the markdown files to land into based on the json
-    provided
+def make_output_dirs(table_names: Iterable[str], base_dir: Optional[str] = None):
```
Comment (iaindillingham) [inline action/find_save_tools.py:33]: If we ask for an `Iterable`, then we can pass in a `list`, `tuple`, `dict` or whatever. Notice that in the associated commit, we haven't updated our tests: these still pass in a `dict`.

Comment (iaindillingham) [reply action/find_save_tools.py:33]: Notice, too, that the Pythonic way of saying "`X` is optional" is `Optional[X]` rather than `Union[X, None]`. (Replacing `str` for whatever type is appropriate.) Whilst [they are equivalent](https://docs.python.org/3.8/library/typing.html#typing.Optional), `Optional` makes our intentions much clearer.

File: action/redaction_tools.py [hunk 4]
```diff
@@ -1,56 +1,22 @@
...
-        small_no_limit (int): the number at which values should be redacted. Defaults to
-            5.
-
-    returns
-        boolean (redaction_needed): True if needs redaction, False if not.
-    """
-    # Convert table into list for iteration
-    all_values = table.values.tolist()
+def contains_small_numbers(table: pd.DataFrame, threshold: int = 5) -> bool:
+    """Does `table` contain numbers less than or equal to `threshold`?"""
```
Comment (iaindillingham) [inline action/redaction_tools.py:7]: If I think I might refactor the function, then I've skimped on the docstring.

File: tests/test_redaction_tools.py [hunk 5]
```diff
@@ -1,55 +1,50 @@
...
-def test_check_for_low_numbers():
-    no_redact_data = {"test_col_1": [10, 20], "test_col_2": [20, 40]}
-    no_redact_df = pd.DataFrame(data=no_redact_data)
 
-    # use check_for_low_numbers() on a dataframe that does not require redaction
-    result = check_for_low_numbers(table=no_redact_df, small_no_limit=5)
-    assert result is False
+class TestContainsSmallNumbers:
+    def test_contains_small_numbers(self):
+        table = pd.DataFrame({"col_1": [5, 6], "col_2": [6, 6]})
```
Comment (iaindillingham) [inline tests/test_redaction_tools.py:11]: Notice that we focus on the boundary condition.

**Comment by iaindillingham** [comment]
I've refactored `find_save_tools.make_output_dirs` (previously `find_save_tools.make_folders`), so this is now ready for PR and hopefully merge. Would you mind taking a brief second look, @CarolineMorton? I've added some comments, which hopefully help.

File: tests/test_redaction_tools.py [hunk 6]
```diff
@@ -1,55 +1,50 @@
...
+            [[6, 6], [6, 6]],
+            index=pd.Index(["F", "M"], name="sex"),
+            columns=pd.Index([0, 1], name="has_condition"),
+        )
 
-    # assert redacted
-    # print("TABLE 2: ", test_table_2)
-    assert test_table_2 == "REDACTED"
+        assert obs_cols == exp_cols
+        testing.assert_frame_equal(obs_table, exp_table)
```
Comment (CarolineMorton) [inline tests/test_redaction_tools.py:50]: great to see this in use. we should use more in other places.

File: action/find_save_tools.py [hunk 7]
```diff
@@ -1,52 +1,46 @@
...
-        os.makedirs(path, exist_ok=True)
-        for folder_name, table_dets in folder_names.items():
-            full_path = os.path.join(path, folder_name)
-            os.makedirs(full_path, exist_ok=True)
+    for table_name in table_names:
+        if base_dir is None:
+            dir_out = table_name
+        else:
+            dir_out = os.path.join(base_dir, table_name)
+        os.makedirs(dir_out, exist_ok=True)
```
Comment (CarolineMorton) [inline action/find_save_tools.py:46]: Good refactoring.

**Review by CarolineMorton — APPROVED** [review]
Looks good to me. Well done for refactoring all that code into something better. Good to see.
